src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs

```diff
@@ -1,48 +1,56 @@
+    private readonly IOcelotLogger _logger;
+
+    public PollyCircuitBreakingDelegatingHandler(
+        DownstreamRoute route,
+        IHttpContextAccessor contextAccessor,
+        IOcelotLoggerFactory loggerFactory)
+    {
+        _route = route;
+        _contextAccessor = contextAccessor;
+        _logger = loggerFactory.CreateLogger<PollyCircuitBreakingDelegatingHandler>();
+    }
+
+    private IPollyQoSProvider<HttpResponseMessage> GetQoSProvider()
     {
-        private readonly IPollyQoSProvider _qoSProvider;
-        private readonly IOcelotLogger _logger;
+        Debug.Assert(_contextAccessor.HttpContext != null, "_contextAccessor.HttpContext != null");
+        return _contextAccessor.HttpContext.RequestServices.GetService<IPollyQoSProvider<HttpResponseMessage>>();
+    }
 
-        public PollyCircuitBreakingDelegatingHandler(
-            IPollyQoSProvider qoSProvider,
-            IOcelotLoggerFactory loggerFactory)
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+        CancellationToken cancellationToken)
+    {
+        var qoSProvider = GetQoSProvider();
+        try
         {
-            _qoSProvider = qoSProvider;
-            _logger = loggerFactory.CreateLogger<PollyCircuitBreakingDelegatingHandler>();
+            // at least one policy (timeout) will be returned
+            // CircuitBreakerAsyncPolicy can't be null
+            // CircuitBreaker constructor will throw if no policy is provided
+            var policy = qoSProvider.GetCircuitBreaker(_route).CircuitBreakerAsyncPolicy;
+            return await policy.ExecuteAsync(async () => await base.SendAsync(request, cancellationToken));
         }
-
-        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        catch (BrokenCircuitException ex)
         {
-            try
-            {
-                var policies = _qoSProvider.CircuitBreaker.Policies;
-                if (!policies.Any())
-                {
-                    return await base.SendAsync(request, cancellationToken);
-                }
-
-                var policy = policies.Length > 1
-                    ? Policy.WrapAsync(policies)
-                    : policies[0];
-
-                return await policy.ExecuteAsync(() => base.SendAsync(request, cancellationToken));
-            }
-            catch (BrokenCircuitException ex)
-            {
-                _logger.LogError("Reached to allowed number of exceptions. Circuit is open", ex);
-                throw;
-            }
-            catch (HttpRequestException ex)
-            {
-                _logger.LogError($"Error in {nameof(PollyCircuitBreakingDelegatingHandler)}.{nameof(SendAsync)}", ex);
-                throw;
-            }
+            _logger.LogError("Reached to allowed number of exceptions. Circuit is open", ex);
+            throw;
```

Comment:
For what purpose do you re-throw? To be caught by Exception-to-Error logic?

Reply:
That's the original logic, I didn't want to mess with it.

Reply:
OK. I hate this Exception-to-Error mapping too. Bad design! We have a lot of problems because of this feature...
Global UnhandledException handler should catch exceptions for logging purposes. Only for rare cases we could catch exception to stabilize the behavior...

Reply:
Imho, it should be all the exceptions should be catched or none, but now it's, sometimes yes, sometimes no, it's not consistent.

Reply:
OK Let's ignore for now...